identity_stardust/src/client/client_ext.rs

```diff
@@ -0,0 +1,484 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use std::borrow::Cow;
+use std::ops::Deref;
+use std::str::FromStr;
+
+use iota_client::api_types::responses::OutputResponse;
+use iota_client::block::address::Address;
+use iota_client::block::output::feature::IssuerFeature;
+use iota_client::block::output::feature::SenderFeature;
+use iota_client::block::output::unlock_condition::AddressUnlockCondition;
+use iota_client::block::output::unlock_condition::GovernorAddressUnlockCondition;
+use iota_client::block::output::unlock_condition::StateControllerAddressUnlockCondition;
+use iota_client::block::output::AliasId;
+use iota_client::block::output::AliasOutput;
+use iota_client::block::output::AliasOutputBuilder;
+use iota_client::block::output::BasicOutputBuilder;
+use iota_client::block::output::Feature;
+use iota_client::block::output::Output;
+use iota_client::block::output::OutputId;
+use iota_client::block::output::RentStructure;
+use iota_client::block::output::UnlockCondition;
+use iota_client::block::payload::transaction::TransactionEssence;
+use iota_client::block::payload::Payload;
+use iota_client::block::Block;
+use iota_client::secret::SecretManager;
+use iota_client::Client;
+
+use crate::error::Result;
+use crate::Error;
+use crate::NetworkName;
+use crate::StardustDID;
+use crate::StardustDocument;
+
+/// An extension trait for a [`Client`] that provides helper functions publication and resolution of
+/// DID documents in Alias Outputs.
+#[async_trait::async_trait]
+pub trait StardustClientExt: Sync {
```

Comment:
iota.rs can still be compiled to Wasm and there are no plans to drop Wasm support itself (they only stopped doing iota.rs Wasm bindings themselves), so the idea was that this trait should "just work" for Wasm as long as we get a `Client` instance. Now that you mention it, the `Sync` requirement here might be a problem...

Reply:
If we don't use `iota.js` in the bindings then we need to expose most of what `iota-client` does anyway, so we may as well revert to wrapping it. Including everything to do with the `AliasOutput` (for setting the amount, controllers etc. which `StardustClientExt` does not do) in order to port it to the Wasm bindings ourselves.

Reply:
We probably want to support iota.js as well as iota.rs's node.js bindings, since the latter comes with Stronghold support. To support code-sharing for the Wasm bindings, We can consider reorganizing the current code. The current trait has these methods:

```rust
#[async_trait::async_trait]
pub trait StardustClientExt: Sync {
  async fn new_did(
    &self,
    address: Address,
    document: StardustDocument,
    rent_structure: Option<RentStructure>,
  ) -> Result<AliasOutput> { ... }

  async fn update_did(&self, document: StardustDocument) -> Result<AliasOutput> { ... }

  async fn deactivate_did_output(&self, secret_manager: &SecretManager, did: &StardustDID) -> Result<()> { ... }

  async fn publish_did_output(
    &self,
    secret_manager: &SecretManager,
    alias_output: AliasOutput,
  ) -> Result<StardustDocument> { ... }

  async fn delete_did_output(&self, secret_manager: &SecretManager, address: Address, did: &StardustDID) -> Result<()> { ... }

  async fn resolve_did(&self, did: &StardustDID) -> Result<StardustDocument> { ... }

  async fn resolve_alias_output(&self, did: &StardustDID) -> Result<AliasOutput> { ... }

  async fn network_name(&self) -> Result<NetworkName> { ... }
}
```

We can split it into two traits, based on whether the method needs a `SecretManager` or not, since it is not serializable or otherwise portable across FFI (Stronghold cannot be compiled to Wasm).

```rust
#[async_trait::async_trait]
pub trait StardustClientExtCrud: StardustIdentityClient + Sync {
  async fn new_did(
    &self,
    address: Address,
    document: StardustDocument,
    rent_structure: Option<RentStructure>,
  ) -> Result<AliasOutput> { ... }

  async fn update_did(&self, document: StardustDocument) -> Result<AliasOutput> { ... }

  async fn resolve_did(&self, did: &StardustDID) -> Result<StardustDocument> { ... }

  async fn resolve_alias_output(&self, did: &StardustDID) -> Result<AliasOutput> { ... }

  async fn network_name(&self) -> Result<NetworkName> { ... }
}

#[async_trait::async_trait]
pub trait StardustIdentityClient: Sync {
  async fn get_network_hrp(&self) -> Result<String>;

  async fn resolve_alias_output(&self, did: &StardustDID) -> Result<(AliasId, OutputId, AliasOutput)>;

  async fn get_rent_structure(&self) -> Result<RentStructure>;

  /* more helper methods that StardustClientExtCrud requires */
}

#[async_trait::async_trait]
impl StardustIdentityClient for Client { ... }

impl<T> StardustClientExtCrud for T where T: StardustIdentityClient {}
```

And the other part:

```rust
// Contains methods that need a SecretManager.
pub trait StardustClientExtPublishing: Sync {
  async fn deactivate_did_output(&self, secret_manager: &SecretManager, did: &StardustDID) -> Result<()> { ... }

  async fn publish_did_output(
    &self,
    secret_manager: &SecretManager,
    alias_output: AliasOutput,
  ) -> Result<StardustDocument> { ... }

  async fn delete_did_output(&self, secret_manager: &SecretManager, address: Address, did: &StardustDID) -> Result<()> { ... }
}
```

(Names temporary).

We should be able to implement `StardustIdentityClient` for `IotaJsClient` and/or `IotaRsNodeJsBindingClient` since all the parameters of the trait's methods are serializable, so we can re-use the code from `StardustClientExtCrud`. `StardustIdentityClient` is meant to be an abstraction over the necessery networking functionality the extension trait needs. `StardustClientExtPublishing` is only meant as an extension trait for the Rust API. Its functionality will have to be reimplemented from scratch for the Wasm binding, due to the lack of `SecretManager` portability.

For convenience in Rust we can implement a trait that looks like the initial `StardustClientExt` and just delegates to both traits.

Does that make sense?

In the interest of getting this PR merged soon, I would not do the refactoring in this one, though.

Reply:
Seems like a good approach. We will pursue this direction for the Wasm bindings once this PR is merged.